4216: Image pull per runtime class KEP

kiashok opened the pull request:
- One-line PR description: Adding new KEP for supporting image pull per runtime class

- Issue link: https://github.com/kubernetes/enhancements/issues/4216

- Other comments: This KEP was discussed in SIG-Windows and Kubelet GC WorkGroup and sig-nodes slack group.

kiashok commented:
cc @jsturtevant @mikebrow @marosset @haircommander @SergeyKanzhelev @aravindhp @dmcgowan @kevpar @dcantah

aravindhp commented:
/sig windows

aravindhp commented:
/ok-to-test

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux,windows, Darwin etc). The platform matcher on linux attempts to check for host and guest OS match and optionally for a variant match as well. However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci Image ...
+
+This problem can be fixed if containerd, CRI and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index. This document details the changes needed to support pulling images per runtime class.
+Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. In windows, differnt versions of a multiarch image could be pulled based on hyperV and process isolation. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
```
```suggestion
Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. In windows, differnt versions of a multiarch image could be pulled based on hyperV and process isolation. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
```

It would be helpful to give a few reasons why this is helpful to add additional context.

kiashok replied:
The motivation section has more details with examples.

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+Currently, the windows platform matcher looks for an exact OS version match between the host and guest while pulling images. This behavior is not accurate for hyperV isolation as the UVM/sandbox OS version could be different from the host OS version.  
+
+For example, a user might want to pull an image like python (a manifest list) on a WS2022 host with an intent of running it in on a WS2019 hyperV isolated UVM. But unfortunately, the current windows platform matcher behavior only forces the WS2022 version of the python image to be pulled (due to the exact version match between host and guest). If a manifest list does not have a version that match ...
+Extending image pulls to specify an optional runtime class would greatly help to extend windows platform matcher behavior depending on whether hyperV or process isolation is being used. Additionally, having support to pull images per runtime class might also help projects like supporting different snapshotters with different runtimes as well.
```
This section is mostly a duplicate of above, having it here is probably enough as the summary gives you the rational and this give more indepth details.

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+With this feature, multiple versions of the same image can exist on a single node. For exmaple, python image can exist on the node for WS2022 (20348 build number) and WS2019(17703 build number) depending on the guestOSVersion mentioned in containerd's toml file for a runtime. Therefore, it is important for kubelet, containerd and other components to identify this. The following sections details t ...
+
+#### ImageSpec
+Include field to specify optional runtime class to use to pull image
```
```suggestion
Include field on CRI `ImageSpec` message to specify optional runtime class to use to pull image
```

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+    // If the runtime handler is unknown, this request should be rejected.  An
+    // empty string should select the default handler, equivalent to the
+    // behavior before this feature was added.
+    <mark>string runtime_handler = 19;</mark>
```
Is this the only change required to also address the `snapshotter per runtimeclass` scenario?

kiashok replied:
No, CRI has another store for storing snapshots related data. I haven't looked into it in detail but this is one of the main files that would need changes - https://github.com/containerd/containerd/blob/main/pkg/cri/store/snapshot/snapshot.go .

This KEP does not intend to cover snapshots per runtime class changes. It is a project I have heard folks are interested in several meetings so I just added it as one of the projects that would be benefitted from this image pull per runtime class. I will call it out in the KEP specifically to avoid any confusion.

marosset replied:
Can we say that snapshots per runtime class changes are out of scope for this KEP and say something like implementation details will be handled by each container runtime?

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+
+### Update containerd.Store interface with options
+
+The fpllowing is done to be able to pass the runtimehandler information while publishing create/update/delete events (as they eventually call the criservice.UpdateImage API that need to know the runtime handler). RuntimeHandler information is not stored in any of core containerd data structures as this is intended to only be a concept at CRI layer.
```
```suggestion
The following is done to be able to pass the runtimehandler information while publishing create/update/delete events (as they eventually call the criservice.UpdateImage API that need to know the runtime handler). RuntimeHandler information is not stored in any of core containerd data structures as this is intended to only be a concept at CRI layer.
```

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+    bool verbose = 2;
+}
+
+message ImageStatusResponse {
```
Where is the runtimeclass being used in the Response?

kiashok replied:
ImageStatusResponse.Image.RuntimeClass could have the runtime class that was used to pull the image. I thought this would be useful for cases where the ImageStatusRequest was made without specifying the runtime class, in which case containerd would have used the default runtimeclass for that platform. Filling this information in the ImageStatusResponse might help to add some info logging to kubelet.

jsturtevant replied:
got it, looks like `Image` has the `ImageSpec` already imbedded into it so it will get the updates.

jsturtevant replied:
https://github.com/kubernetes/kubernetes/blob/6f5fa2eb2f4dc731243b00f7e781e95589b5621f/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1412C24-L1412C24

kiashok replied:
yeah

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtime class:
```
Does kublet currently have access to the required information at every point when making the requests below?

kiashok replied:
Yes. This is needed primarily:
1. EnsureImageExists() which is used to ensure an image exists before starting a pod or container. This has the needed info to identify the runtime class.
I see the same for RemoveImageRequest, ImageStatus and GC

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
```
For Windows there are a few tests we can run, Are there any that would validate this on the Linux side?

kiashok replied:
Yes we should have tests that test for functionality and regression across different platforms. Will call this out in the testing section.

jsturtevant reviewed (commented):
Thanks for the detailed KEP.  Once minor suggestion is to multiple and multi-line sentences with return character to make commenting on individual sentences easier.

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+know that this has succeeded?
+-->
+This KEP describes the detailed design and major API changes needed in kubelet, CRI and containerd to support image pull with optional runtime class option.
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
```
snpashotter per runtimeclass is not in scope. It is one of the projects which folks have been talking about addressing for some time now. That needs some additional changes and should be addressed separately. But it will benefit from this work as image will already be addresses by (name, runtimeClass). I mentioned it in the summary/motivation section to only point out snapshotter per runtime class when it happens will benefit from these changes. Should I explicitly call it out as a out of scope/non-goals?

kiashok replied:
@aravindhp  to answer your question "Are changes to tooling like ctr and crictl in or out of scope?" 
No, they are in scope and I do intend to make the necessary changes for these. It is discussed in the later part of this KEP.

jsturtevant replied:
> Should I explicitly call it out as a out of scope/non-goals?

yes please,  Has there been some discussion or preliminary work done to make sure these changes are compatible?

kiashok replied:
@jsturtevant I made changes to KEP to call it out.
Also, I don't think is any active effort that I know of for supporting snapshotter per runtime class. But we briefly discussed this is the kubelet garbage collection WG with @mikebrow  and @haircommander .

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+
+1. Before running a container, kuebelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
+
+PullImageRequest has a field for ImageSpec which can hold info for the runtime class used to pull this image:
```
If nothing is filled in, the ImagePullRequest will have an empty field for runtime class and containerd would check this and use default runtime handler for the platform if none was specified. It would need changes in pkg/kubelet/images/puller.go pullImage() function  and EnsureImageExists() function in pkg/kubelet/images/image_manager.go. They do have access to the ImageSpec and the runtime handlers need to be set by the called. In short, there are changes that will be needed.

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+}
+</pre>
+
+3. Image cleanup/kubelet garbage collection: Currently, kubelet GC will call into RemoveImage() on containerd side to remove unused images. kubelet will only pass the image name and containerd will currently delete all instances of the image. This needs to be extended on kubelet to ask for deleting (imageName, runtimeclass) to be more efficient. Optionally, we could add a behavior on containerd s ...
```
> > This needs to be extended on kubelet to ask for deleting (imageName, runtimeclass) to be more efficient.
> 
> Is this possible?

I think this is GC behavior that would need to account for this once we change image to be referred by (imageName, runtimeclass) and I think this should be possible as runtimeclass field is added to the ImageSpec. If none is specified, containerd would try to delete the image with default runtime class.
cc @haircommander as he is working on altering kubelet GC behavior. I can work with you before changing this.

> 
> > Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image is that is was kubelet GC wants to do.
> 
> I don't follow this sentence. What happens if all the images are cleaned up but some are still being used?

If some image is still being used, then containerd should not be successful in deleting the images, right? I think my sentence should be reworded as follows: 
"Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image so that kubelet GC has a way to delete all versions of an image if it has identified that the images are not being used by any container/pod."

Does this answer the questions? Or am I missing smth?

jsturtevant replied:
I am not super familiar with this area, but It seems like there could be some tricky edge cases here, if kubelet thinks it deleted an image but containerd doesn't and vice versa but I don't really know enough

kiashok replied:
I think that problem already exists with kubelet and containerd behavior for image deletion - didn't mean to indicate any exisitng behavior would change. Reworded this to reduce confusion.

aravindhp commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+</pre>
+
+
+### Windows platform matcher optional changes
```
What is `Windows platform matcher`?

kiashok replied:
Each platform has its own platform matchers defined in https://github.com/containerd/containerd/tree/main/platforms/default*.go files. Windows platform matcher is defined in https://github.com/containerd/containerd/tree/main/platforms/default_windows.go . I will add link to this in KEP for better clarity too.

kiashok replied:
Added more details for this to the section

aravindhp reviewed (commented):
@kiashok thanks for opening the KEP. I suggest removing the template notes so that the KEP is easier to review.  I also suggest wrapping lines to a reasonable length so that we can site specific areas and also allow future diffs to be small.

I have also left a few comments.

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
+
+## Proposal
+
```
I do not see a comment box for your previous comment so I am addressing both here:

"Are changes to tooling like ctr and crictl in or out of scope?"

I intend to make necessary changes needed to crictl and ctr to support this feature. Have listed out the crictl changes that will be made in the later parts of the KEP. 

"I suggest expanding on your motivation section here and clearly calling out what the desired outcome is."

I think the summary and motivation sections are kinda duplicates so I have broken them into two parts with examples and the desired outcome in detail. Not sure if I have anything additional to add. Do you have something in mind that I need to add that I could be missing?

aravindhp replied:
The summary and motivation are not really duplicates. Think of summary as a sort of release note for the feature. The motivation section goes into why this feature is important for users and leads into goals and non-goals.

kiashok replied:
Okay. But I think summary needs to have a little of motivation anyway to completely be able to give a summary of what this KEP is trying to achieve.
Either ways, I am not sure what more detail is needed in either of these sections. I have put in all the details I could think about. If there is smth you see is missing please let me know

kiashok replied:
I added another example to motivation section as that was all I think could be added to these two sections.

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1338 @@
+2. RemoveImageRequest
+3. ImageStatusRequest
+
+### Containerd changes:
```
We do not intend to address any other runtimes. We could have someoene from other runtimes like CRI-O to add or support this. 

cc @mikebrow @haircommander  any thoughts here?

haircommander replied:
> It almost feels like this KEP should be specific to the k8s components. 

yeah I agree here. I don't think the containerd specific changes are relevant here

kiashok replied:
@haircommander should the KEP just had the kubelet changes needed? Can I completely remove the containerd specifics then?  
@mikebrow @dcantah @dmcgowan what would be a good place to have the containerd changes? Should it just be a doc that I have linked to my PR?

kiashok commented:
> Thanks for the detailed KEP. Once minor suggestion is to multiple and multi-line sentences with return character to make commenting on individual sentences easier.

Hi James, thanks for the review! I've addressed them and also broken longer sentences/multi line sentences with new line characters wherever I could.

kiashok commented:
> @kiashok thanks for opening the KEP. I suggest removing the template notes so that the KEP is easier to review. I also suggest wrapping lines to a reasonable length so that we can site specific areas and also allow future diffs to be small.
> 
> I have also left a few comments.

Hi @aravindhp , thanks for taking a look! I've addressed the comments. Also broke down longer sentences/multi line sentences with new line characters wherever I could.

haircommander commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1222 @@
+### CRI/kubelet changes:
+
+With this feature, multiple versions of the same image can exist on a single node.
+For exmaple, python image can exist on the node for WS2022 (20348 build number) and WS2019(17703 build number) depending on the guestOSVersion mentioned in containerd's toml file for a runtime.
```
example*

haircommander commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,1222 @@
+2. Image status request/response can also make use of the newly added runtime class field in ImageSpec to specify the runtime class being used for the image. If none is specified, containerd will assume that the request is for the default runtime class defined for that platform in containerd.toml
+
+<pre>
+message ImageStatusRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Verbose indicates whether to return extra information about the image.
+    bool verbose = 2;
```
can you use diff syntax in this section so it's easier to see what's being added?

kiashok replied:
There are no new changes being added in this message. I've highlighted ImageSpec field to indicate that this is where the new runtime class field is being added and will be populated. The last paragraph in the section write above this talks about how ImageSpec is being extended to add runtime handler field.

haircommander commented on keps/sig-node/4216-image-pull-per-runtime-class/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+status: provisional
+creation-date: 2023-09-15
+reviewers:
+  - TBD
```
you can add me here for review from CRI-O side

haircommander reviewed (changes requested):
A couple of nits. I agree with @aravindhp that the containerd specific changes aren't as relevant. I think this should cover kubelet + cri changes, and then how the specific cri implementations wish to handle it is up to them

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+## Summary
+
+
+Today, container images in containerd are identified by the image name/id. Images can either be a manifest or manifest list/index. Containerd's platform matcher code kicks in for index images and is responsible for pulling an appropriate manifest from the index of that image.
```
```suggestion
Today, container images in containerd are identified by the image name/id. Images can either be a manifest or manifest list/index. When an image is specified platform matching logic in containerd is used to pull the appropriate manifest from the index of that image
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+In containerd, each platform has its own platform matcher which is used to
+decide which manifest to pull from an image index/manifest list.
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux,windows, Darwin etc).
```
```suggestion
Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux, Windows, Darwin, etc).
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux,windows, Darwin etc).
+The platform matcher on linux attempts to check for host and guest OS match and optionally for a variant match as well.
+However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci ImageSpec is primarily used only for windows manifests) and if a manifest with such an exact match is found, it is pulled. Otherwise an error is thrown. <br> This is the desired behavior for process isolated windows containers as its shares the kernel with the host OS. ...
+For example, we could have a UVM  running Windows Server 2019 image on a Windows Server 2022 host and this should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
```
```suggestion
For example, Windows Server 2019 container images can run inside a Utility VM (UVM) on a Windows Server 2022 host and because of this the image pulls should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+` Note: ` UVM is short for utility VM. It is a light weight VM that boots a special version of Windows that is pupose built to do just one thing- run containers. UVMs provide strong isolation boundaries and do not share kernel with the host OS. They are transparent to the user and do not require any management from the user.
+
+This problem can be fixed if containerd, CRI and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index.
```
```suggestion
This problem can be fixed if containerd, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index.
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+` Note: ` UVM is short for utility VM. It is a light weight VM that boots a special version of Windows that is pupose built to do just one thing- run containers. UVMs provide strong isolation boundaries and do not share kernel with the host OS. They are transparent to the user and do not require any management from the user.
+
+This problem can be fixed if containerd, CRI and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index.
```
Here you mention that we can solve the problem if we include runtimeClass when specifying the image.
Can you breifly eplain how this will get solved?
Is the plan to have the platform matcher check the runtimeClass config for the target and match the OS version to the one specified instead of the host's OS version?

kiashok replied:
Yes. We would extend containerd behavior to have a platform matcher per runtime class - If guestPlatform has been specified for a hyperV runtime class in containerd.toml, then the platform matcher for this runtime classs should be based on the guestPlatform.OSVersion rather than the host's OS version.

I had this detail in the KEP before but got rid of it as it was talking about containerd specific implementation and got feedback that we should have only k8s changes in this KEP. Will add a sentence or two to this paragraph as an example of how containerd behavior can be changed.

kiashok replied:
Added the following to this paragraph:
"An example implementation in the containerd container runtime on how the pull implementation can be extended to support this is as follows:
On containerd, the windows platform matcher is responsible for pulling an appropriate manifest from an image index on windows platforms. Today, this platform matcher on windows always looks for a manifest with the OSVersion that matches the host and pulls it. This can be changed to pick a manifest with OSVersion that matches the one set for a particular runtime class in containerd.toml instead of the host OSVersion. "

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+This document details the changes needed to support pulling images per runtime class.
+Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
+For windows in particular, different versions of a multiarch image could be pulled based on hyperV and process isolation and we could remove the necessity for exact host and guest OS version matches for hyperV isolation.
```
Maybe add a coment saying this could be useful for any VM based container solution?

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+This KEP attempts to goes over the details of the most significant changes needed in containerd, CRI, kubelet and tools like ctr and crictl to support this feature.
+
+## Motivation
```
A lot of the details here seem redundant with above details. Maybe you can say see above for the example?

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+### Goals
+
+This KEP describes the detailed design and major API changes needed in kubelet, CRI and containerd to support image pull with optional runtime class option.
```
```suggestion
This KEP describes the detailed design and API changes needed in kubelet, CRI, and containerd to support image pull with optional runtime class option.
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+### CRI/kubelet changes:
+
+With this feature, multiple versions of the same image can exist on a single node.
+For exmaple, python image can exist on the node for WS2022 (20348 build number) and WS2019(17703 build number) depending on the guestOSVersion mentioned in containerd's toml file for a runtime.
```
```suggestion
For exmaple, python image can exist on the node for WS2022 (OSVersion = 20348) and WS2019(OsVersion = 17703) depending on the guestOSVersion mentioned in containerd's toml file for a runtime.
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+With this feature, multiple versions of the same image can exist on a single node.
+For exmaple, python image can exist on the node for WS2022 (20348 build number) and WS2019(17703 build number) depending on the guestOSVersion mentioned in containerd's toml file for a runtime.
+Therefore, it is important for kubelet, containerd and other components to identify this.
```
```suggestion
Therefore, it is important for kubelet, containerd, and other components to identify all of the OSversion's supported by the host.
```
Feel free to reword but maybe be more specific then 'this'

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
```
```suggestion
Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the `Image` struct in CRI to add a field for runtimeClass will make it easier to extend Kubelet API to include this information. 
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,792 @@
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
```
Let's fill out the Alpha graduation criteria.

Something like
- Above CRI changes are implemented
- Kubelet changes to pass runtimeClass on CRI calls implmented behind a feature flag

marosset commented:
Can you add a prr yaml file for this KEP?
https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/template/nnnn.yaml

marosset commented:
/label lead-opted-in

kiashok commented:
> A couple of nits. I agree with @aravindhp that the containerd specific changes aren't as relevant. I think this should cover kubelet + cri changes, and then how the specific cri implementations wish to handle it is up to them

Removed the containerd specific changes from this KEP and added a note to the non-goals section that implementation details for the different container runtimes should be handled by the runtimes themselves. Thanks!

kiashok commented:
> Can you add a prr yaml file for this KEP? https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/template/nnnn.yaml

I added this. Who should the reviewers be? I have it as "TBD" for now.

kiashok commented:
> cc @jsturtevant @mikebrow @marosset @haircommander @SergeyKanzhelev @aravindhp @dmcgowan @kevpar @dcantah

Hi all, I have addressed all the review comments. Could you please take another look and if it looks ok? Thanks.

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,800 @@
+## Summary
+
+
+Today, container images in containerd are identified by the image name/id. Images can either be a manifest or manifest list/index. When an image is specified, platform matching logic in containerd is used to pull the appropriate manifest from the index of that image.
```
I will change this intro to just refer to runtime. But in a later section I have used containerd as an example where I feel its necessary to understand the end to end motivation and what I am proposing to make sense

aravindhp replied:
It is fine to call out containerd as an example. But we don't need to fully go through the containerd specific implementation details as part of this KEP.

kiashok replied:
Fixed this and updated the PR with the changes. Containerd is only mentioned as an example in 2-3 places but in general I have changed to "container runtime"

aravindhp commented on keps/sig-node/4216-image-pull-per-runtime-class/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+title: Image pull per runtime class
+kep-number: 4126
```
The KEP number needs to match. This should be `4216`.

kiashok commented:
> cc @jsturtevant @mikebrow @marosset @haircommander @SergeyKanzhelev @aravindhp @dmcgowan @kevpar @dcantah

All the tests are passing. Could you please take a look?

haircommander commented:
LGTM

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+status: implementable
+creation-date: 2023-09-15
+reviewers:
+  - TBD
```
```suggestion
```

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,807 @@
+2. RemoveImageRequest
+3. ImageStatusRequest
+
+### Kubelet changes
```
Can you specify the name of the feature-gate that will be used to control having the kubelet extra information described below here and also update it in the kep.yaml file?

marosset commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,808 @@
+- Regression testing to ensure exisiting functionality is not regressed when no runtime class is specified while pulling the image
+- Functional/integration tests for CRI, and kubelet.
+
+[ ] I/we understand the owners of the involved components may require updates to
```
check this box please

aravindhp commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,677 @@
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
```
Please add check marks where appropriate

aravindhp commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,677 @@
+
+#### GA
+
+- No conrete plan for GA yet
```
`concrete`.

I suggest saying we will make a decision on when to GA based on the beta feedback.

aravindhp commented on keps/sig-node/4216-image-pull-per-runtime-class/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.29"
+  beta: "v1.yy"
+  stable: "v1.zz"
```
Please fill in best case release versions here

jsturtevant commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+` Note: ` The kubelet changes will be under a feature-gate called `RuntimeClassInCRICalls`
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtimeClass:
+
+1. Before running a container, kuebelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
```
```suggestion
1. Before running a container, kubelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
```

jsturtevant commented:
LGTM, for sig windows

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. All the code changes needed will be under feature gate `RuntimeClassInCriCalls` that will be off by default for alpha.
+Default exisiting behavior of kubelet will be ensured.
```
If I enable this, and then I end up pulling the same image name/tag but for different runtime classes to a node, then I disable this feature, will kubelet get confused about what image to use for a new container with that image name/tag?

Similarly, what happens to running containers? I guess they will be fine, even if they restart? But if the the pod is rescheduled to a node with flag off then of course it would get the host-matched image, but that seems fine.

Ideally, we should be able to disable this without any extra clean up needed. But if we do need some cleanup, we need to document it here.

kiashok replied:
If the different image versions of the same image exist, turning off feature-gate without cleanup step will cause problems. This would require a cleanup step. Adding in these details as well.

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+Default exisiting behavior of kubelet will be ensured.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+It should be a no-op
```
Similar questions to above - what if there are images for different runtime classes on kubelet and you re-enable this, what happens on container restart, pod restart, maybe kubelet / node restart? Are we in the clear for all of these?

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+Exisiting behavior of kubelet should not change when runtime class is not speified when image is pulled.
```
s/Exisiting/Existing/

there are a bunch of spelling errors / typos throughout the KEP. You may want to run it through a markdown spell checker. There are a few out there, I have used this one and it's kinda slow but works well enough: https://hub.docker.com/r/tmaier/markdown-spellcheck/

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Image pulls for non-default runtime class goes through. Alternately, one could check the status of the feature gate.
```
So, the spirit of this question is "I am an operator with 10,000 clusters, and I have this feature enabled. But this feature is causing problems. Who will I break if I turn off this feature?". Thus, the idea is for a sort of global view of when it is in use. We get a little bit of it from the feature gate. But ideally what we have is a metric - say, for a given node how many times a different image was pulled than would have been otherwise (or even better if instrumentation is OK with the volume, would be to be able to identify the specific workloads).

Note this is a beta question, not required for alpha.

kiashok replied:
An easy way to check this is how many images have a non-default runtime handler associated with it. This metric can be added for beta for sure. Updated the doc with the same.

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+###### How can someone using this feature know that it is working for their instance?
+
+Image pull goes through successfully when non-default runtimehandler is used.
+For example, if host is a windows node, user should be able to pull images for hyperV isolation even if the container image OSVersion does not match the host OS version.
```
Is there, say, a `kubectl` command that could tell you it worked, and pulled the right image?

kiashok replied:
Hmm not that I am aware of. We could add metrics in beta that would help find out how many images are using a non-default runtime class when pulling images.

johnbelamaric replied:
thanks that would be good

johnbelamaric commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+
+###### Does this feature depend on any specific services running in the cluster?
+
+kubelet
```
No need for this. The spirit of this question is to identify dependencies - for example, an autoscaler likely requires a metrics service.

johnbelamaric reviewed (changes requested):
Ok, looks pretty good. I do have some questions around whether any clean up is needed after disablement.

johnbelamaric reviewed (commented):
Few items to follow up on later, but PRR looks good. Will approve after the SIG approval is in.

jsturtevant commented:
/approve 
for sig-windows

/assign @haircommander @SergeyKanzhelev @mrunalp
for sig-node

mrunalp commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+
+This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
+An example implementation in the containerd container runtime on how the pull implementation can be extended to support this is as follows:
+On containerd, the windows platform matcher is responsible for pulling an appropriate manifest from an image index on windows platforms. Today, this platform matcher on windows always looks for a manifest with the OSVersion that matches the host and pulls it. This can be changed to pick a manifest with OSVersion that matches the one set for a particular runtime class in containerd.toml instead of ...
```
Can we add an example of what a runtime class for this looks like?

mrunalp commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,611 @@
+
+1. PullImageRequest
+2. RemoveImageRequest
+3. ImageStatusRequest
```
ListImages

mrunalp commented on keps/sig-node/4216-image-pull-per-runtime-class/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates:
+  - name: RuntimeClassInCRICalls
```
I suggest making this more specific.

kiashok replied:
I changed the feature gate name to 'RuntimeClassInImageCriApi' in the yaml and readme. Let me know if this name sounds better. Thanks

kiashok commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+On k8s, the pod and runtime class yamls would look as follows:
+
+<pre>
+apiVersion: node.k8s.io/v1
```
Yes we would need a runtimeclass is OSVersion is different. Do you mean we should pass it in like this?
apiVersion: v1 
kind: Pod 
metadata: 
  name: mypod 
spec:
  containers: 
  - name: busybox 
    image: docker.io/library/busybox:1.35.0-musl 
    runtimeClassName: runhcs-wcow-runhcs

jsturtevant replied:
@mrunalp could you explain a bit more what you mean? Setting the `runtimeClassName` on the pod spec is how this works today. 

There is some documentation on the various uses cases here https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1301-windows-runtime-class#story-2---forward-compatibility-with-hyper-v

mrunalp replied:
In the example above, OSVersion is `10.0.20348`. If there are multiple different OSVersions that need to be supported, mutliple RuntimeClasses will have to be created.

kiashok replied:
We want this to be more generic for all platforms. For windows, this would be OSVersion but linux does not use OSVersion and if they need to override default behavior, they could override the variant information. The guestPlatform field mentioned in the example for containerd is of type ocispec.Platform . Passing osversion alone will not be generic enough

kiashok replied:
container runtimes will use runtimeclasses to read the associated config from the toml and set up platform matchers (example in containerd) as needed.

mrunalp replied:
Thanks!

mikebrow replied:
Either way works for me... tuple or triplet  (image reference/digest, runtime class, platform (the oci platform struct)).  We can have more runtime classes one per unique desired platform (and name them accordingly), or pass the portions of the platform struct that the user desires to switch on, from the image spec portion(s) of the pod spec... Do all of the containers need to be in the same vm/type?

kiashok replied:
@mikebrow "Do all of the containers need to be in the same vm/type?" not sure I understand this. All the containers inside of a pod have to be using the same runtime class. If a pod is started as hyperV isolated all the containers inside the pod would be hyperV isolated as well.

I think I would prefer having runtime class to pass this information for reasons mentioned by James.

aravindhp commented:
> Few items to follow up on later, but PRR looks good. Will approve after the SIG approval is in.

@johnbelamaric SIG-node has approved

johnbelamaric commented:
/approve
/lgtm

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+## Summary
+
+
+Today, container images in container runtimes are identified by the image name/id. Images can either be a manifest or manifest list/index. When an image is specified, platform matching logic in containerd is used to pull the appropriate manifest from the index of that image.
```
```suggestion
Today, container images in container runtimes are identified by the image name (aka. image reference) and/or image digest (usually a SHA256 cryptographic hash of the content). The content of the image referenced is either an image manifest or an image index manifest. The image index is typically arranged to contain a list of image manifests by platform (linux, windows,...).  When the reference is to an image index, platform matching logic is used to identify and pull the appropriately matching image manifest from the index for the platform that the container is being asked to run on.
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+}
+</pre>
+
+For example, in containerd runtime , each platform has its own platform matcher which is used to
```
```suggestion
In this image index example, each unique platform definition can be matched to the version of the image built for that specific platform, affording the container runtime the ability to
```

mikebrow replied:
sure.. just suggesting here to add another sentence talking about the platform definition.. before introducing the platform matcher.. for that definition..

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+
+For example, in containerd runtime , each platform has its own platform matcher which is used to
+decide which manifest to pull from an image index/manifest list.
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux, Windows, Darwin, etc).
```
```suggestion
For example, containerd's default platform matcher will automatically select the image manifest to use from an image index based on the digest to the image manifest for the OS/platform where the node is running (Linux, Windows, Darwin, etc).
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+decide which manifest to pull from an image index/manifest list.
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux, Windows, Darwin, etc).
+The platform matcher on linux attempts to check for host and guest OS match and optionally for a variant match as well.
+However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci ImageSpec is primarily used only for windows manifests) and if a manifest with such an exact match is found, it is pulled. Otherwise an error is thrown. <br> This is the desired behavior for process isolated windows containers as its shares the kernel with the host OS. ...
```
```suggestion
However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci ImageSpec is primarily used only for windows manifests) and if a manifest with such an exact match is found, it is pulled. Otherwise an error is thrown. <br> This is the desired behavior for process isolated windows containers running on the same kernel with the host OS. However, it is not ideal for hyperV isolated windows containers. This is because OS version of the host and the UVM image need not be the same.
```

kiashok replied:
it does not on the same kernel as the OS. The windows base images have a kernel whose version is the same as the host OS kernel version.

mikebrow replied:
isolated windows containers running on the same version of the kernel that the host is running

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux, Windows, Darwin, etc).
+The platform matcher on linux attempts to check for host and guest OS match and optionally for a variant match as well.
+However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci ImageSpec is primarily used only for windows manifests) and if a manifest with such an exact match is found, it is pulled. Otherwise an error is thrown. <br> This is the desired behavior for process isolated windows containers as its shares the kernel with the host OS. ...
+For example, Windows Server 2019 container images can run inside a Utility VM (UVM) on a Windows Server 2022 host and because of this the image pulls should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os ...
```
```suggestion
For example, Windows Server 2019 container images can run inside a Utility VM (UVM) on a Windows Server 2022 host and because the container will be running in the UVM, the platform matching image pulls for the UVM instances should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+
+` Note: ` UVM is short for utility VM. It is a light weight VM that boots a special version of Windows that is purpose built to do just one thing- run containers. UVMs provide strong isolation boundaries and do not share kernel with the host OS. They are transparent to the user and do not require any management from the user.
+
+This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
```
```suggestion
This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/imageDigest, runtimeClass) instead of the imageName/imageDigest. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+` Note: ` UVM is short for utility VM. It is a light weight VM that boots a special version of Windows that is purpose built to do just one thing- run containers. UVMs provide strong isolation boundaries and do not share kernel with the host OS. They are transparent to the user and do not require any management from the user.
+
+This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
+An example implementation in the containerd container runtime on how the pull implementation can be extended to support this is as follows:
```
```suggestion
An example implementation how the pull implementation in containerd can be extended to support this is as follows:
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+
+This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
+An example implementation in the containerd container runtime on how the pull implementation can be extended to support this is as follows:
+On containerd, the windows platform matcher is responsible for pulling an appropriate manifest from an image index on windows platforms. Today, this platform matcher on windows always looks for a manifest with the OSVersion that matches the host and pulls it. This can be changed to pick a manifest with OSVersion that matches the one set for a particular runtime class in containerd.toml instead of ...
```
```suggestion
On containerd, the windows platform matcher is responsible for pulling an appropriate image manifest from an image index on windows platforms. Today, this platform matcher on windows always looks for an image manifest with the OSVersion that matches the host and pulls it. This can be changed to pick the image manifest with the platform OSVersion that matches the one set for a particular runtime class in containerd.toml instead of the host OSVersion.
```

mikebrow commented on keps/sig-node/4216-image-pull-per-runtime-class/README.md:
```diff
@@ -0,0 +1,662 @@
+### Goals
+
+This KEP describes the detailed design and API changes needed in kubelet and CRI to support image pull with optional runtime class option.
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
```
** if we add platform to containerd runtime class....and change the platform matcher in containerd.. then the user/admin can still pick on a "node" basis what the behavior will be for said default/existing runtime classes.

kiashok replied:
Hmm not sure I understand the concern here :) Are you suggesting this be added as smth the KEP could achieve?

mikebrow replied:
not really a concern.. this is a more a mitigation/statement regarding what could be done "if user does not use the runtime class option with image pull" in regards to the goal stated here "to keep the functionality same as it is today."

mikebrow reviewed (commented):
see comments.. then based on decision points.. can update rest of text to align
